README.md
# et_secrets [![Build Status](https://travis-ci.org/evertrue/et_secrets-cookbook.svg)](https://travis-ci.org/evertrue/et_secrets-cookbook)

Handles the installation and setup of various Evertrue-specific stuff related to secrets and Vault.

# Recipes

## default

Just a wrapper. Installs Vault and Consul client

## renew_vault_token

Renew the token at `data_bags: vault/tokens/#{ENV}/vault/default`

## replace_vault_token

Replace the token at `data_bags: vault/tokens/#{ENV}/vault/default` when its TTL reaches 3600 seconds. Possibly a complete replacement for `renew_vault_token`.

## replace_vault_token

Replace the token at `data_bags: vault/tokens/#{ENV}/vault/default` when its TTL reaches 3600 seconds. Possibly a complete replacement for `renew_vault_token`.

Short Description

**USE ONLY FOR TESTING!!!**

Completely re-initializes the vault at `http://localhost:8200`, generates a new root token, and saves it to `"#{Chef::Config[:file_cache_path]}/test-kitchen_root_token"`. Also generates a client token (for testing) and saves it to `"#{Chef::Config[:file_cache_path]}/test-kitchen_test_token"`. **ALL EXISTING VAULT DATA WILL BE ERASED!!!**

## renew_vault_token

This recipe renews the Vault token contained in the `vault/tokens` data bag item by calling `/auth/tokens/renew-self` on the Vault API. It requires `node['etc_environment']['VAULT_ADDR']` to be set.

## vault_init

Used only during testing. Re-initializes the vault to a like-new state, unseals it and populates `node['et_secrets']['test_token']` with a test token with default privileges.

Never (**ever, ever**) put this recipe in the run list on a real server!!!

# Resources

## replace_vault_token

```ruby
replace_vault_token 'default' do
  vault_host "http://vault.service.#{node.chef_environment}-us-east-1.consul:8200"
  data_bag_name 'vault'
  data_bag_item_name 'tokens'
  options(
    ttl: '8670h',
    policies: %w(default apps)
  )
  top_level_key node.chef_environment
  min_remaining_ttl 3600
  accessor_token data_bag_name('vault', 'tokens')[node.chef_environment]['vault']['worker_token']
end
```

#### Properties

* **name** - The name associated with the token in the data bag and in the Vault server metadata.
* **vault_host** - Self explanatory. E.g. `http://vault.host:8200`
* **options** - A hash passed verbatim to the Vault token-create API. E.g. `{ policies: ['default', 'apps'] }`
* **token_to_replace** - A string containing the actual token to replace. If provided, the resource will verify this token against the Vault API instead of trying to use a token from the data bag. If a new token is generated (because the specified token is expired or invalid) it will still be stored in the specified data bag.
* **accessor_token** - The token to use for processing the request (See **Required Permissions**)
* **min_remaining_ttl** - If the existing token has less than this many seconds remaining on its TTL, it will be replaced. If unspecified, a new token will be generated each time the resource is run.
* **data_bag_name** and **data_bag_item_name** - Where to find and store the token.
* **top_level_key** - The top level key for the path within the data bag for the token that is to be be replaced. E.g.
```json
{
  "id": "your_data_bag",
  "top_level_key": {
    "vault": {
      "default": "YOUR_OLD_KEY"
    }
  }
}
```

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests with `kitchen test`, ensuring they all pass
6. Submit a Pull Request using Github

## License and Authors

Author:: EverTrue (devops@evertrue.com)


CHANGELOG.md
# et_secrets CHANGELOG

This is the Changelog for the et_secrets cookbook.

## [Unreleased][unreleased]

### Changes

### Fixes

## v4.0.1 - (2016-09-30)

### Fixes

* Do not allow automatic reboots (moved from the `et_consul` cookbook where it didn't belong)

## v4.0.0 - (2016-08-16)

### Breaking

* Bump et_consul to 4.0

## v3.0.0 - (2016-08-16)

### Breaking

* Pin hashicorp-vault cookbook at exactly v2.4.0 (Includes [some breaking changes](https://www.vaultproject.io/docs/install/upgrade-to-0.6.html))

### Changes

* Set default lease ttl to 1 year
* Don't pin specific vault version (Allows Vault version to rise to 0.6.0)

## v2.3.2 - (2016-07-07)

### Changes

* Test that encrypted data bag insertion occurred with an actual encrypted data bag

### Fixes

* Encrypted data bag save: Convert source data bag item to hash earlier in the process

## v2.3.1 - (2016-07-07)

### Fixes

* Fix saving of encrypted data bag

## v2.3.0 - (2016-07-07)

### Changes

* Alert PD before token expires

## v2.2.2 - (2016-06-06)

### Fixes

* Allow replacement token name to be different than replace_vault_token resource name
* Move vault tokens to a separate data bag location to deal with lack of source control

## v2.2.1 - (2016-06-03)

### Fixes

* Replace data_bag_name/data_bag_item

## v2.2.0 - (2016-06-03)

### Changes

* New recipe and resource: replace_vault_token
* Set global max and default lease TTLs

## v2.1.2 - (2016-05-24)

### Fixes

* Pass ALL of the arguments to fetch on redirect

## v2.1.1 - (2016-05-24)

### Fixes

* Handle redirects from vault server
* No illegal characters in name tags

## v2.1.0 - (2016-05-24)

### Changes

* New recipe: `renew_vault_token`
* Make vault_init a regular recipe (rather than a test cookbook) so that other cookbooks can use it

### Fixes

* Include et_consul dependency (already in use in et_secrets::default)
* Use sftp transport for test kitchen
* Use "mocking" instead of "bootstrap_expect" to set bootstrap expect number for consul

## v2.0.1 - (2016-04-14)

### Fixes

* Drop specifying `address` in Vault Consul Service definition
    - Fixes how the Consul DNS works and provides the correct private IP for Vault

## v2.0.0 - (2016-04-04)

### Changes

* Migrate Vault code from et_consul back into here
* Upgrade to Vault 0.5.2, which is breaking due to various changes
* Use `hashicorp-vault ~> 2.1`

## v1.0.0 - (2015-09-03)

### Changes

* Bumping to v1.0.0 to reflect introduction to prod
* Add Vault policies as JSON files, though they are not yet used by this cookbook for anything

## v0.2.0 - (2015-08-25)

### Changes

* Switch to using as-yet-unreleased version of the official `hashicorp-vault` cookbook, instead of our fork (which likely will wither & die)
* Add integration test for `advertise_addr`

## v0.1.1 - (2015-08-25)

### Fixes

* Set Vault to listen on `0.0.0.0:8200`, allowing outside traffic

## v0.1.0 - (2015-08-25)

### Added

* Initial release


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

RSpec.configure do |config|
  config.formatter = :documentation
  config.color = true
  config.platform = 'ubuntu'
  config.version = '14.04'
end

at_exit { ChefSpec::Coverage.report! }


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: et_secrets
# Spec:: default
#
# Copyright (c) 2015 EverTrue, All Rights Reserved.

require 'spec_helper'

describe 'et_secrets::default' do
  context 'when all attributes are default, on Ubuntu 14.04' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new
      runner.converge(described_recipe)
    end
    it 'converges successfully' do
      chef_run # This should not raise an error
    end
  end
end


test/integration/cookbooks/test_deps/metadata.rb
name 'test_deps'
version '1.0.1'


recipes/vault_init.rb
ruby_block 'clear vault data store' do
  block do
    # Clear the Vault data store (from Consul)
    uri = URI('http://localhost:8500/v1/kv/vault?recurse')
    req = Net::HTTP::Delete.new uri
    Net::HTTP.start(uri.hostname, uri.port) { |http| http.request req }
  end
end

execute 'restart vault' do
  command 'service vault restart'
end

ruby_block 're-initialize and unseal vault' do
  block do
    sleep 5

    # Re-initialize the vault
    http = Net::HTTP.new 'localhost', 8200
    req = Net::HTTP::Put.new '/v1/sys/init'
    req.body = { secret_shares: 1, secret_threshold: 1 }.to_json
    response = http.request req

    unless response.code.to_i == 200
      raise "ERROR: Got response #{response.code} from vault during " \
            "re-initialization. Body: #{response.body}"
    end

    parsed_response = JSON.parse(response.body)
    unseal_key = parsed_response['keys'].first
    root_token = parsed_response['root_token']

    File.open("#{Chef::Config[:file_cache_path]}/test-kitchen_root_token", 'w') do |f|
      f.write root_token
    end

    # Unseal the vault
    req = Net::HTTP::Put.new '/v1/sys/unseal'
    req.body = { key: unseal_key }.to_json
    response = http.request req

    unless response.code.to_i == 200
      raise "ERROR: Got response #{response.code} from vault during " \
            "unsealing. Body: #{response.body}"
    end

    sleep 5

    req = Net::HTTP::Post.new '/v1/auth/token/create'
    req['X-Vault-Token'] = root_token
    req.body = { display_name: 'test', policies: %w(default), ttl: '1000h' }.to_json
    response = http.request req

    unless response.code.to_i == 200
      raise "ERROR: Got response #{response.code} from vault during " \
            "test token generation.\n" \
            "Root token: #{root_token}\n" \
            "Body: #{response.body}"
    end

    parsed_response = JSON.parse(response.body)

    unless parsed_response['auth']['lease_duration'] == 3_600_000
      raise "Test token TTL was not long enough: #{parsed_response['auth']['lease_duration']}"
    end

    File.open("#{Chef::Config[:file_cache_path]}/test-kitchen_test_token", 'w') do |f|
      f.write parsed_response['auth']['client_token']
    end
  end
end


recipes/renew_vault_token.rb
# The first value here should only actually be used during testing
vault_token = data_bag_item('vault', 'tokens')[node.chef_environment]['vault']['default']

ruby_block 'renew vault token' do
  block do
    begin
      require 'net/http'

      def fetch(uri_str, vault_token, limit = 10)
        raise ArgumentError, 'too many HTTP redirects' if limit == 0

        uri = URI uri_str
        req = Net::HTTP::Post.new uri
        req['X-Vault-Token'] = vault_token
        req.body = { increment: 2_592_000 }.to_json
        req.content_type = 'application/json'
        response = Net::HTTP.start(uri.hostname, uri.port) { |http| http.request req }

        case response
        when Net::HTTPSuccess then response
        when Net::HTTPRedirection then fetch(response['location'], vault_token, limit - 1)
        else
          Chef::Log.error "Bad response from vault server (#{uri_str}): #{response.code}\n" \
                          "Token: #{vault_token}\n" \
                          "Body: #{response.body}\n"
          response.error!
        end
      end

      # If this file exists, assume we're actually testing and use the token
      # from there instead. This is a hack to get around the order of attribute
      # writing operations.
      if File.exist? "#{Chef::Config[:file_cache_path]}/test-kitchen_test_token"
        vault_token = File.read "#{Chef::Config[:file_cache_path]}/test-kitchen_test_token"
      end

      vault_addr = node['etc_environment']['VAULT_ADDR']

      response = fetch "#{vault_addr}/v1/auth/token/renew-self", vault_token

      if JSON.parse(response.body)['auth']['client_token'] != vault_token
        # Hammond pointed out that the protocol leaves room for the client_token
        # to change upon renewal. Our setup doesn't handle that, so in the event
        # that it does occur, we bail.
        raise "HALP: client_token received from #{vault_addr} does not match " \
              'what we sent.'
      end
    rescue => e
      Chef::Log.error "Error renewing token with #{vault_addr}: #{e.message}"
      raise e
    end
  end
end


resources/replace_vault_token.rb
#
# Cookbook Name:: et_secrets
# Resource:: replace_vault_token
#
# Copyright (c) 2016 EverTrue, All Rights Reserved.

resource_name :replace_vault_token
default_action :renew

property :token_name,               kind_of: String, name_attribute: true
# A full URL is expected for the vault host.
# E.g.
# http://localhost:8200
property :vault_host,
         kind_of: String,
         required: true,
         callbacks: { 'Must be a valid URL' => proc { |p| valid_url p } }
property :data_bag_name,      kind_of: String, required: true
property :data_bag_item_name, kind_of: String, required: true
property :options,            default: {}

# Manually specify the actual token to replace. Usually this is automatically
# gathered from the data bag key mentioned below, but this will override that
# value.
property :token_to_replace, kind_of: String

# The privileged token that will be used to request the new token
# requires the "create" capability on the "auth/token/create" path
property :accessor_token, kind_of: String, required: true

# Enable Pagerduty alerting when token is about to expire
property :pagerduty_key, kind_of: String

# The top level key for the path within the data bag for the token that
# is to be be replaced.

# e.g.
# {
#   "id": "your_data_bag",
#   "top_level_key": {
#     "vault": {
#       "default": "YOUR_OLD_KEY"
#     }
#   }
# }
property :top_level_key, kind_of: String, required: true

# If the existing token has more than this many seconds remaining before
# it expires, renewal will not take place. If unset, the token will be
# replaced every time.
property :min_remaining_ttl, kind_of: Integer

def self.valid_url(url)
  url =~ /\A#{URI.regexp(%w(http https))}\z/
end

def vault
  @vault ||= begin
    require 'vault'
    require 'pagerduty' if pagerduty_key

    v = Vault::Client.new address: vault_host, token: accessor_token
    if (ttl = v.auth_token.lookup_self.data[:ttl]) < (3_600 * 24 * 3) && ttl != 0
      Chef::Log.warn "The accessor token for #{token_name} expires in #{ttl} seconds"
      alert_token_expiring ttl if pagerduty_key
    end
    v
  end
end

def alert_token_expiring(ttl)
  msg = "The accessor token used for replacing #{token_name} expires in less than 3 days. " \
    'Please replace this token with a new one using the token/create API.'
  pd = Pagerduty.new pagerduty_key
  pd.trigger(
    msg,
    incident_key: "accessor token for #{token_name} expiring soon",
    client: node.name,
    details: {
      ttl: ttl
    }
  )
end

def new_token
  vault.auth_token.create options
end

def secret
  @secret ||= begin
    key_paths = %w(
      /etc/chef/encrypted_data_bag_secret
      /tmp/kitchen/encrypted_data_bag_secret
    )
    key_path = key_paths.find { |kp| ::File.exist? kp }
    ::File.read(key_path).chomp
  end
end

def data_bag_save(item_json)
  encrypted_new_item = Chef::EncryptedDataBagItem.encrypt_data_bag_item(item_json, secret)
  new_item = Chef::DataBagItem.from_hash(encrypted_new_item)
  new_item.data_bag(data_bag_name)
  new_item.save
end

action :renew do
  chef_gem 'vault'
  chef_gem 'pagerduty' if pagerduty_key
  db = data_bag_item(data_bag_name, data_bag_item_name).to_hash
  begin
    token_string = token_to_replace || db[top_level_key]['vault'][token_name]
    if token_string
      Chef::Log.debug 'Existing token found in data bags. Looking it up in Vault.'
      existing_token = vault.auth_token.lookup token_string
    end
  rescue Vault::HTTPClientError => e
    # Bad token indicates the token does not exist, which is fine for us; We'll
    # just set a new one.
    raise e unless e.message =~ /\* bad token/
    Chef::Log.warn 'Existing token is bad.'
  end

  # If:
  # - there is no existing token OR
  # - there's no minimum TTL OR
  # - there is a minimum TTL AND the token is below it
  # ... get a new token.
  if !existing_token ||
     !property_is_set?(:min_remaining_ttl) ||
     property_is_set?(:min_remaining_ttl) &&
     existing_token.data[:ttl] < min_remaining_ttl
    Chef::Log.debug(
      'Token does not exist, is invalid, or has expired. Requesting a new ' \
      "one with name #{token_name} and options: #{options.inspect}."
    )
    begin
      db[top_level_key]['vault'][token_name] = new_token.auth.client_token
    rescue => e
      Chef::Log.fatal "Value of auth: #{new_token.auth.inspect}"
      raise e
    end
    data_bag_save db
    new_resource.updated_by_last_action true
  else
    Chef::Log.debug 'Token exists and is not expired. Doing nothing.'
  end
end
